docs/design/daemon.md

```diff
@@ -0,0 +1,118 @@
+
+## Functionality
+
+The DaemonSet supports standard API features:
+- create
+  - The spec for DaemonSets has a pod template field.
+  - Using the pod’s nodeSelector field, DaemonSets can be restricted to operate over nodes that have a certain label. For example, suppose that in a cluster some nodes are labeled ‘app=database’. You can use a DaemonSet to launch a datastore pod on exactly those nodes labeled ‘app=database’.
+  - Using the pod's node name field, DaemonSets can be restricted to operate on a specified nodeName.
+  - The PodTemplateSpec used by the DaemonSet is the same as the PodTemplateSpec usedby the Replication Controller.
+  - We will not guarantee that daemon pods show up on nodes before regular pods - run ordering is out of scope for this abstraction in the initial implementation.
+  - The initial implementation of DaemonSet does not guarantee that Daemon pods show up on nodes (for example because of resource limitations of the node), but makes a best effort to launch Daemon pods (like Replication Controllers do with pods). Subsequent revisions might ensure that Daemon pods show up on nodes, preempting other pods if necessary.
+  - The DaemonSet controller adds an annotation "kubernetes.io/created-by: \<json API object reference\>"
+  - YAML example:
+```YAML
+  apiVersion: v1
+  kind: Daemon
+  metadata:
+    labels:
+      app: datastore
+    name: datastore
+  spec:
+    template:
+      metadata:
+        labels:
+          app: datastore-shard
+      spec:
+        nodeSelector: 
+          app: datastore-node
+        containers:
+          name: datastore-shard
+          image: kubernetes/sharded
+          ports:
+            - containerPort: 9042
+              name: main
+```
+  - commands that get info
+    - get (e.g. kubectl get daemonsets)
+    - describe
+  - Modifiers
+    - delete (if --cascade=true, then first the client turns down all the pods controlled by the DaemonSet (by setting the nodeName to a non-existant name); then it deletes the DaemonSet; then it deletes the pods)
+    - label
+    - update (only allowed to selector and to nodeSelector and nodeName of pod template)
+    - DaemonSets have labels, so you could, for example, list all DaemonSets with a certain label (the same way you would for a Replication Controller).
+  - In general, for all the supported features like get, describe, update, etc, the DaemonSet works in a similar way to the Replication Controller. However, note that the DaemonSet and the Replication Controller are different constructs.
+
+### Persisting Pods
+  - Ordinary livenes probes specified in the pod template work to keep pods created by a DaemonSet running.
+  - If a daemon pod is killed or stopped, the DaemonSet will create a new replica of the daemon pod on the node.
+
+### Cluster Mutations
+  - When a new node is added to the cluster the DaemonSet starts the daemon on the node (if the node’s labels match the user-specified selectors).
```

Comment:
When a new node is added to the cluster, the DaemonSet controller starts daemon pods on the node for DaemonSets whose pod template nodeSelectors match the node’s labels.